File: src/main/java/net/masterthought/dlanguage/stubs/types/DLanguageImportDeclarationStubElementType.java
package net.masterthought.dlanguage.stubs.types;

import com.intellij.lang.ASTNode;
import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import net.masterthought.dlanguage.psi.DLanguageImport;
import net.masterthought.dlanguage.psi.impl.DLanguageImportImpl;
import net.masterthought.dlanguage.stubs.DLanguageImportDeclarationStub;
import net.masterthought.dlanguage.utils.DUtil;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

/**
 * Created by francis on 3/15/2017.
 */
public class DLanguageImportDeclarationStubElementType extends DNamedStubElementType<DLanguageImportDeclarationStub, DLanguageImport> {
    public DLanguageImportDeclarationStubElementType(String debugName) {
        super(debugName);
    }

    @Override
    public DLanguageImport createPsi(@NotNull DLanguageImportDeclarationStub stub) {
        return new DLanguageImportImpl(stub, this);
    }

    @Override
    public boolean shouldCreateStub(ASTNode node) {
        return DUtil.definitionNode(node);
    }

    @NotNull
    @Override
    public DLanguageImportDeclarationStub createStub(@NotNull DLanguageImport psi, StubElement parentStub) {
        return new DLanguageImportDeclarationStub(parentStub, this, psi.getName());
    }

    @Override
    public void serialize(@NotNull DLanguageImportDeclarationStub stub, @NotNull StubOutputStream dataStream) throws IOException {
        dataStream.writeName(stub.getName());
    }

    @NotNull
    @Override
    public DLanguageImportDeclarationStub deserialize(@NotNull StubInputStream dataStream, StubElement parentStub) throws IOException {
        return new DLanguageImportDeclarationStub(parentStub, this, dataStream.readName());
    }
}

File: dlang/psi-impl/src/main/kotlin/io/github/intellij/dlanguage/stubs/types/DLanguageImportDeclarationStubElementType.kt
package io.github.intellij.dlanguage.stubs.types

import com.intellij.psi.PsiElement
import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import io.github.intellij.dlanguage.psi.DLanguageImportDeclaration
import io.github.intellij.dlanguage.psi.impl.DLanguageImportDeclarationImpl
import io.github.intellij.dlanguage.psi.interfaces.DVisibility
import io.github.intellij.dlanguage.psi.interfaces.DVisibilityStubKind
import io.github.intellij.dlanguage.psi.interfaces.stubKind
import io.github.intellij.dlanguage.stubs.DLanguageImportDeclarationStub

class DLanguageImportDeclarationStubElementType(debugName: String) : DStubElementType<DLanguageImportDeclarationStub, DLanguageImportDeclaration>(debugName) {
    override fun createPsi(stub: DLanguageImportDeclarationStub): DLanguageImportDeclaration {
        return DLanguageImportDeclarationImpl(stub, this)
    }

    override fun createStub(
        psi: DLanguageImportDeclaration,
        parentStub: StubElement<out PsiElement?>
    ): DLanguageImportDeclarationStub {
        return DLanguageImportDeclarationStub(
            parentStub,
            this,
            psi.visibility().stubKind,
            (psi.visibility() as? DVisibility.Package)?.name)
    }

    override fun serialize(stub: DLanguageImportDeclarationStub, dataStream: StubOutputStream) {
        dataStream.writeByte(stub.visibility.ordinal)
        if (stub.visibility == DVisibilityStubKind.PACKAGE_SPECIFIC)
            dataStream.writeUTF(stub.visibility.name)
    }

    override fun deserialize(dataStream: StubInputStream, parentStub: StubElement<out PsiElement?>): DLanguageImportDeclarationStub {
        val visibilityKind = enumValues<DVisibilityStubKind>()[dataStream.readUnsignedByte()]

        val packageNameRestriction = if (visibilityKind == DVisibilityStubKind.PACKAGE_SPECIFIC)
            dataStream.readUTF()
        else
            null
        return DLanguageImportDeclarationStub(
            parentStub,
            this,
            visibilityKind,
            packageNameRestriction
        )
    }
}
